gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ImmunizationMapper.java

```diff
@@ -0,0 +1,301 @@
+
+        ehrCompositionList.forEach(ehrComposition -> {
+            var immunizationObservationStatements = getImmunizationObservationStatements(ehrComposition);
+
+            immunizationObservationStatements
+                .forEach(observationStatement -> {
+                    var mappedImmunization = mapImmunization(ehrComposition, observationStatement, patientResource,
+                        encounterList);
+                    mappedImmunizationResources.add(mappedImmunization);
+                });
+        });
+
+        return mappedImmunizationResources;
+    }
+
+    private Immunization mapImmunization(RCMRMT030101UK04EhrComposition ehrComposition,
+        RCMRMT030101UK04ObservationStatement observationStatement,
+        Patient patientResource, List<Encounter> encounterList) {
+
+        ImmunizationMapperParameters immunizationMapperParameters = new ImmunizationMapperParameters();
+
+        var id = observationStatement.getId().getRoot();
+        var identifier = getIdentifier(id);
+        var note = buildNote(observationStatement);
+        var date = getObservationDate(observationStatement);
+
+        Extension recordedTimeExtension = getRecordedTimeExtension(ehrComposition);
+        Extension vaccineExtension = createVaccineProcedureExtension(observationStatement);
+
+        var practitioner = ParticipantReferenceUtil.getParticipantReference(observationStatement.getParticipant(), ehrComposition);
+        var encounter = getEncounterReference(encounterList, ehrComposition.getId());
+
+        immunizationMapperParameters.setIdParam(id);
+        immunizationMapperParameters.setIdentifierParam(identifier);
+        immunizationMapperParameters.setNoteParam(note);
+        immunizationMapperParameters.setDateParam(date);
+        immunizationMapperParameters.setRecordedTimeExtensionParam(recordedTimeExtension);
+        immunizationMapperParameters.setVaccineExtensionParam(vaccineExtension);
+        immunizationMapperParameters.setPatientParam(new Reference(patientResource.getIdElement()));
+        immunizationMapperParameters.setEncounterParam(encounter);
+        immunizationMapperParameters.setPractitionerParam(practitioner);
+
+        return createImmunization(immunizationMapperParameters);
+    }
+
+    private Reference getEncounterReference(List<Encounter> encounterList, II ehrCompositionId) {
+        if (ehrCompositionId != null) {
+            var matchingEncounter = encounterList.stream()
+                .filter(encounter -> matchingIds(encounter.getId(), ehrCompositionId))
+                .findFirst();
+
+            if (matchingEncounter.isPresent()) {
+                return new Reference(matchingEncounter.get().getIdElement());
+            }
+        }
+
+        return null;
+    }
+
+    private boolean matchingIds(String encounterId, II ehrCompositionId) {
+        return encounterId.equals(ehrCompositionId.getRoot());
+    }
```

Comment:
Just a favour to ask with this one, get in contact with Bart C on this, i know he is moving it to a Util class so depending on how merges first, if you do then just ask him to change this to instead use his UtilClass please!

Reply:
> Just a favour to ask with this one, get in contact with Bart C on this, i know he is moving it to a Util class so depending on how merges first, if you do then just ask him to change this to instead use his UtilClass please!

Bart, got back to me and said he no longer needs this in his class has there is a bit of work around it so i am good to keep this in, thanks